sql/opt: Invalidate memo for queries using RLS when user changes

spilchen opened the pull request:
When reusing a query from the statement cache, we generally allow reuse if the query is to be executed by a different user. However, this approach is problematic for queries that involve row-level security (RLS), as the executing user directly influences the injected expressions. To ensure correctness, this considers the memo as stale when the user changes if RLS was used in the query.

Epic: CRDB-11724
Release note: None

Closes #136717

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

One thing I'm just realizing (maybe you've already thought of this) is that UDFs and procedures can use the `SECURITY DEFINER` setting which invokes the routine with the privileges of the user that defined it, not the user invoking it. I would assume that RLS policies for the definer are then applied to table accesses within the routine, but I haven't checked what Postgres does here. I would also assume this creates some additional complexity for  memo staleness—we know that the definer of the routine will never change (unless a descriptor version changes which will invalidate the cached memo anyway), but we need to be sure that we aren't accidentally tracking the user definer in rlsMeta and confusing it as the current user.

---

As I reviewed this, I had an idea for a different implementation of RLS that would both address this memo staleness problem and prevent any RLS details from leaking into the optimizer logic. I'm not confident it's worth it to change course at this point, but I thought I'd lay it out regardless—if we don't want to pursue it now maybe we will in the future.

I think RLS policies could be implemented entirely in the opt catalog as views. When the optimizer comes across a data source reference, e.g., `t` in `SELECT * FROM t`, it asks the opt catalog to resolve `t`. If `t` is a table in which the current user has an RLS policy, then it is resolved as a view with that policy as a filter, rather than a table. In the context of an insert or update, the opt catalog would resolve as a table with addition check constraints. There might be some things to iron out, e.g., would the RLS view/table have the same StableID? The nice thing about this approach is that the optimizer doesn't need to be made aware of RLS at all—RLS is sort of a syntactic sugar on top of the standard set of SQL features implemented by the optimizer. Additionally, making sure that the cached memo isn't incorrectly used is a matter of simply making sure that `optView.Equals` and `optTable.Equals` return false if RLS was used and the username is different. 

With an implementation like this, RLS starts to look a lot like a special form of updatable views. 

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-:-OHx_LE-B2mvbc-CTTsK:b-nw31qz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @spilchen)

___
*[`pkg/sql/opt/row_level_security.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHxbZAf3i8_7Bw8VK29:-OHxbZAf3i8_7Bw8VK2A:bfxz6d3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38bd80633a572e4d9b00668a9549e9689d750edd/pkg/sql/opt/row_level_security.go#L30)):*
> ```Go
> }
> 
> func (r *RowLevelSecurityMeta) MaybeInit(evalCtx *eval.Context, hasAdminRole bool) {
> ```

nit: I'd prefer to minimize the API by replacing the `*eval.Context` parameter with `username.SQLUsername`.
___
*[`pkg/sql/opt/row_level_security.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHxasUp00-zS7nEn9V_:-OHxasUp00-zS7nEn9Va:bs2ib96) ([raw file](https://github.com/cockroachdb/cockroach/blob/38bd80633a572e4d9b00668a9549e9689d750edd/pkg/sql/opt/row_level_security.go#L41)):*
> ```Go
> // Clear unsets the initialized property. This is used as a test helper.
> func (r *RowLevelSecurityMeta) Clear() {
> 	r.IsInitialized = false
> ```

Shouldn't this clear `User` and `HasAdminRole` too?
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHxb-qB11AQDPbopztb:-OHxb-qB11AQDPbopztc:b9nx76u) ([raw file](https://github.com/cockroachdb/cockroach/blob/38bd80633a572e4d9b00668a9549e9689d750edd/pkg/sql/opt/optbuilder/row_level_security.go#L38)):*
> ```Go
> 		return
> 	}
> 	b.factory.Metadata().SetRLSEnabled(b.evalCtx, false /* isAdmin */)
> ```

nit: move this above the `if` block and pass `isAdmin` as the second argument


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
I need to double-check, but in postgres, when calling a UDF or procedure, RLS policies are applied based on the invoker, not the definer, even if `SECURITY DEFINER` is set. I’ll confirm this and verify that CRDB behaves the same way. This is a good point to highlight.

I like your idea of implementing RLS as a form of updatable view. However, does CRDB currently support updatable views? I was experimenting with it and kept encountering `"ERROR: 'view1' is not a table"`. I’d prefer to keep the current approach for now but would like to discuss this further. Maybe we could implement the read path like a view and introduce the write path once updatable views are supported.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-:-OHy9acD-v-QF7rJ1KpK:b-u1s7qz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/row_level_security.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHxasUp00-zS7nEn9V_:-OHy9l965tvdcLlfG-Ob:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/38bd80633a572e4d9b00668a9549e9689d750edd/pkg/sql/opt/row_level_security.go#L41)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Shouldn't this clear `User` and `HasAdminRole` too?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 8 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-:-OHyOOiKA_YZYybzEAP3:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @spilchen)

___
*[`pkg/sql/opt/metadata.go` line 1297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHyLnFI6_YA3bYhCobF:-OHyLnFI6_YA3bYhCobG:b-1c0w1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/38bd80633a572e4d9b00668a9549e9689d750edd/pkg/sql/opt/metadata.go#L1297)):*
> ```Go
> 	// RLS policies.
> 	if hasAdminRole, err := optCatalog.HasAdminRole(ctx); err != nil {
> 		return false, err
> ```

Note: if it's possible to get some sort of "object does not exist" error here, consider whether we need to swallow that type of error. Otherwise, you can see queries start to fail because of a cached plan referencing an object that no longer exists, when we really should just invalidate the cached plan.
___
*[`pkg/sql/opt/metadata.go` line 1283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHyMivp6qguYVFbycXr:-OHyMivp6qguYVFbycXs:b26u2dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/76bbf0d4983d86a26a9f859d0c9f88e1fa829e59/pkg/sql/opt/metadata.go#L1283)):*
> ```Go
> 	// rlsMeta is lazily updated. If we didn't initialize it, then we didn't come
> 	// across any RLS enabled tables. So, from a rls point of view the memo is up
> 	// to date.
> ```

The table descriptor version will be incremented if a table had RLS enabled after the query was cached, right?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/opt: Invalidate memo for queries using RLS when user changes

When reusing a query from the statement cache, we generally allow
reuse if the query is to be executed by a different user. However, this
approach is problematic for queries that involve row-level security
(RLS), as the executing user directly influences the injected
expressions. To ensure correctness, this considers the memo as stale
when the user changes if RLS was used in the query.

Epic: CRDB-11724
Release note: None

Closes #136717

spilchen reviewed (commented):
> I need to double-check, but in postgres, when calling a UDF or procedure, RLS policies are applied based on the invoker, not the definer, even if `SECURITY DEFINER` is set. I’ll confirm this and verify that CRDB behaves the same way. This is a good point to highlight.

I was mistaken about this. After testing in postgres, I confirmed that when applying RLS policies, it uses the definer's role for functions declared with `SECURITY DEFINER`. I've updated the code accordingly and added a test case to verify this behavior.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-:-OI6bKxR0qZXvfY9DJ2T:b4d2d9x)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/metadata.go` line 1297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHyLnFI6_YA3bYhCobF:-OI5bF5t0Pg6MiUtqokK:b-okngu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/38bd80633a572e4d9b00668a9549e9689d750edd/pkg/sql/opt/metadata.go#L1297)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Note: if it's possible to get some sort of "object does not exist" error here, consider whether we need to swallow that type of error. Otherwise, you can see queries start to fail because of a cached plan referencing an object that no longer exists, when we really should just invalidate the cached plan.
</blockquote></details>

We shouldn't reach this point if the table has been dropped. Since we already have a pointer to the `cat.Table` object, the lookup would have failed if the table no longer existed. I’ve updated the test to verify this behavior.
___
*[`pkg/sql/opt/metadata.go` line 1283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140069#-OHyMivp6qguYVFbycXr:-OI5aQxt1a7AxaKdVHwZ:b-fb8deb) ([raw file](https://github.com/cockroachdb/cockroach/blob/76bbf0d4983d86a26a9f859d0c9f88e1fa829e59/pkg/sql/opt/metadata.go#L1283)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

The table descriptor version will be incremented if a table had RLS enabled after the query was cached, right?
</blockquote></details>

Yes, that's correct. I updated the test to validate this.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
